src/browser/application.ts

```diff
@@ -0,0 +1,128 @@
+import { Logger } from "pino";
+import { v4 as uuid } from "uuid";
+import { IApplication } from "../base/application";
+import { IInputHandler, IInputMessage } from "../base/input-handler";
+import { ISignalProvider } from "../base/signal-provider";
+import { IStreamProvider } from "../base/stream-provider";
+import { IWebrtcProvider } from "../base/webrtc-provider";
+
+/**
+ * Application constructor options
+ */
+export interface IApplicationOpts {
+  /**
+   * A logger
+   */
+  logger: Logger;
+
+  /**
+   * The name of the capture window
+   */
+  captureWindowTitle: string;
+
+  /**
+   * The ice servers to use
+   */
+  iceServers: RTCIceServer[];
+
+  /**
+   * The signal provider to use
+   */
+  signalProvider: ISignalProvider;
+
+  /**
+   * The stream provider to use
+   */
+  streamProvider: IStreamProvider;
+
+  /**
+   * The webrtc provider to use
+   */
+  webrtcProvider: IWebrtcProvider;
+
+  /**
+   * The input handler to use
+   */
+  inputHandler: IInputHandler;
+}
+
+/**
+ * A browser application - orchestrates the streamer experience
+ */
+export class Application implements IApplication {
+  private opts: IApplicationOpts;
+
+  /**
+   * Default ctor
+   * @param opts ctor options
+   */
+  constructor(opts: IApplicationOpts) {
+    this.opts = opts;
+  }
+
+  /**
+   * Internal boot helper
+   */
+  public async boot() {
+    const {
+      logger,
+      captureWindowTitle,
+      iceServers,
+      signalProvider,
+      streamProvider,
+      webrtcProvider,
+      inputHandler } = this.opts;
+    logger.info("Browser: initializing application");
+
+    const rawDevices = await streamProvider.enumerateDevices();
+    const matchedDevice = rawDevices.find((e) => e.name === captureWindowTitle);
```

Comment:
Could we move this line to before `if (matchedDevice)`? We don't need to start finding the matched device if there is no device.

Reply:
sure, good idea.